File: src/store/global.ts
import { emit as emitStateEvent, listen } from "@tauri-apps/api/event";
import { create } from "zustand";

import {
  GLOBAL_STORE_EVENTS,
  GLOBAL_STORE_METHODS,
  type GlobalState,
} from "@/constants/global";

const useGlobalStore = create<GlobalState>(() => ({
  zw: 0,
  zwIncrement: () => {
    emit(GLOBAL_STORE_METHODS.ZwIncrement);
  },
}));

function emit(method: GLOBAL_STORE_METHODS) {
  emitStateEvent(GLOBAL_STORE_EVENTS.ChangeEvent, {
    method,
  });
}

async function subscribeGlobalStoreSync() {
  const unsubscribeGlobalStoreSyncListener = await listen(
    GLOBAL_STORE_EVENTS.SyncEvent,
    (event) => {
      console.log("LOL");
      const payload = event.payload as {
        key: keyof GlobalState;
        value: GlobalState[keyof GlobalState];
      };
      useGlobalStore.setState({ [payload.key]: payload.value });
    },
  );
  return async () => {
    unsubscribeGlobalStoreSyncListener();
  };
}

export { useGlobalStore, subscribeGlobalStoreSync };

File: src-tauri/src/store/global.rs
use crate::constants::global::{
    events::GlobalStoreEvents,
    keys::GlobalStoreKeys,
    methods::GlobalStoreMethods,
    payload::{GlobalPayload, GlobalPayloadZw},
};
use std::sync::Mutex;
use tauri::{AppHandle, Manager};

pub struct GlobalState {
    pub zw: Mutex<i16>,
}

impl Default for GlobalState {
    fn default() -> Self {
        Self { zw: Mutex::new(0) }
    }
}

pub fn emit_global_sync_event(payload: GlobalPayload, app_handle: AppHandle) {
    let state = app_handle.state::<GlobalState>();

    match payload.method {
        GlobalStoreMethods::ZwIncrement => {
            let mut zw = state.zw.lock().unwrap();
            *zw += 1;

            app_handle
                .emit_all(
                    GlobalStoreEvents::SyncEvent.to_string().as_str(),
                    GlobalPayloadZw {
                        value: *zw,
                        key: GlobalStoreKeys::Zw,
                    },
                )
                .unwrap();
        }
    }
}

pub fn listen_global_change_event(app_handle: AppHandle) {
    let app_handle_clone = app_handle.clone();
    let _id = app_handle.listen_global(GlobalStoreEvents::ChangeEvent.to_string(), move |evt| {
        if let Some(payload) = evt.payload() {
            let data: Result<GlobalPayload, _> = serde_json::from_str(payload);
            match data {
                Ok(data) => emit_global_sync_event(data, app_handle_clone.clone()),
                Err(e) => println!("Failed to deserialize payload: {}", e),
            }
        }
    });
}

